### ISSUE TITLE
Issue with last version

### ISSUE THREAD

**Issue by gusdie** [apexad/homebridge-eufy-robovac#21]
Hi, I’m having some issues in hoobs using the last version. The log is showing this. I have identified the device id and the local key (but not sure if they are ok)

Thanks!

10/7/2020 16:05:14 (node:6631) UnhandledPromiseRejectionWarning: Error: find() timed out. Is the device powered on and the ID or IP correct?
10/7/2020 16:05:14 at /home/hoobs/.hoobs/node_modules/tuyapi/index.js:619:13
10/7/2020 16:05:15 at Timeout._onTimeout (/home/hoobs/.hoobs/node_modules/p-timeout/index.js:25:13)
10/7/2020 16:05:15 at listOnTimeout (internal/timers.js:549:17)
10/7/2020 16:05:15 at processTimers (internal/timers.js:492:7)
10/7/2020 16:05:15 (node:6631) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
10/7/2020 16:05:15 (node:6631) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by apexad** [comment]
I’ve seen some other reports of timeout issues. What vacuum do you have?

**Comment by heyventrilo** [comment]
Same error, I have the 30c Max model.


Here is a print of the issue:

```
UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 7)
(node:16527) 
UnhandledPromiseRejectionWarning: Error: connection timed out
at Socket.client.setTimeout (/usr/lib/node_modules/homebridge-eufy-robovac/node_modules/tuyapi/index.js:302:18)
at Object.onceWrapper (events.js:276:13)
at Socket.emit (events.js:188:13)
at Socket._onTimeout (net.js:445:8)
at listOnTimeout (timers.js:324:15)
at processTimers (timers.js:268:5)
(node:16527)
```

I think is something not from this plugin, is from tuyapi side, but would be great if this plugin handle error exceptions of tuyapi, if not its crashing homebridge.

**Comment by radiomurf** [comment]
I am also seeing these errors. 
`
Jul 21 01:22:21 hoobs homebridge[674]: (node:674) UnhandledPromiseRejectionWarning: Error: connection timed out
Jul 21 01:22:21 hoobs homebridge[674]:     at Socket.<anonymous> (/usr/local/lib/node_modules/homebridge-eufy-robovJul 21 01:22:21 hoobs homebridge[674]:     at Object.onceWrapper (events.js:416:28)
Jul 21 01:22:21 hoobs homebridge[674]:     at Socket.emit (events.js:310:20)
Jul 21 01:22:21 hoobs homebridge[674]:     at Socket._onTimeout (net.js:479:8)
Jul 21 01:22:21 hoobs homebridge[674]:     at listOnTimeout (internal/timers.js:549:17)
Jul 21 01:22:21 hoobs homebridge[674]:     at processTimers (internal/timers.js:492:7)
Jul 21 01:22:21 hoobs homebridge[674]: (node:674) UnhandledPromiseRejectionWarning: Unhandled promise rejection. ThJul 21 01:22:22 hoobs homebridge[674]: Robovac Error {}
`

**Comment by apexad** [comment]
Any once seeing these errors with the 15 model? Seems like it is isolated to the 30 Max.

**Comment by gusdie** [comment]
I have same issues with 11c model

**Comment by heyventrilo** [comment]
I made some manual modifications and i came to this conclussion:

- The package `eufy-robovac` doesn't provide a timeout to `tuyapi`, so default is 5s, so I changed to .2s (this makes a big difference)
- I implemented some `try` and `catch` for error handling
- I added default `callback(true)` to `catch` errors and avoid homebridge to crash. Apple Home always spect a callback(), and if you don't provide a default, it will crash.

That removed the warnings on the console and avoided homebridge to crash.

After a few test scenarios:

- Turn off vacuum / Start homebridge / Turn on vacuum
- Turn on vacuum / Start homebridge / Turn off the vacuum
- Start homebridge / Turn off the vacuum / Turn off vacuum

I was still getting not expected behaviours and crashes on apple home (homebridge was running ok).

So... whats the conclusion ? Having a process inside homebridge that provides timeouts will not work. So I moved to `Homebridge Cmdswitch2` (that I already used) and I created three node scripts based on `eufy-robovac`:

1. vacuum_status: Check vacuum status and write in in a log file
2. vacuum_on: Turn on the vacuum via eufy-robovac and write in in a log file
3. vacuum_off: Turn ff the vacuum via eufy-robovac and write in in a log file

The final result is more reliable and `Homebridge Cmdswitch2`  will check a log file for the vacuum status and if the process crash, doesn't affect anything. 

If someone is interested (and I have time) I can create a documentation to explain how to setup the vacuum directly.

**Comment by gusdie** [comment]
I’m interested if you can do it! Thank you!!!

**Comment by apexad** [comment]
> I made some manual modifications and i came to this conclussion:
> 
> 
> 
> - The package `eufy-robovac` doesn't provide a timeout to `tuyapi`, so default is 5s, so I changed to .2s (this makes a big difference)
> 
> - I implemented some `try` and `catch` for error handling
> 
> - I added default `callback(true)` to `catch` errors and avoid homebridge to crash. Apple Home always spect a callback(), and if you don't provide a default, it will crash.
> 
> 
> 
> That removed the warnings on the console and avoided homebridge to crash.
> 
> 
> 
> After a few test scenarios:
> 
> 
> 
> - Turn off vacuum / Start homebridge / Turn on vacuum
> 
> - Turn on vacuum / Start homebridge / Turn off the vacuum
> 
> - Start homebridge / Turn off the vacuum / Turn off vacuum
> 
> 
> 
> I was still getting not expected behaviours and crashes on apple home (homebridge was running ok).
> 
> 
> 
> So... whats the conclusion ? Having a process inside homebridge that provides timeouts will not work. So I moved to `Homebridge Cmdswitch2` (that I already used) and I created three node scripts based on `eufy-robovac`:
> 
> 
> 
> 1. vacuum_status: Check vacuum status and write in in a log file
> 
> 2. vacuum_on: Turn on the vacuum via eufy-robovac and write in in a log file
> 
> 3. vacuum_off: Turn ff the vacuum via eufy-robovac and write in in a log file
> 
> 
> 
> The final result is more reliable and `Homebridge Cmdswitch2`  will check a log file for the vacuum status and if the process crash, doesn't affect anything. 
> 
> 
> 
> If someone is interested (and I have time) I can create a documentation to explain how to setup the vacuum directly.

@heyventrilo Mind doing a fork of this repo and committing your changes and creating a Pull Request with your changes?  Specifically the timeout addition and try catch block stuff you mentioned.  I want to add that to this plugin.  The other key to this I think is having anytime an error does occur, for the `setup()` function to be ran again.

**Comment by mateen110110** [comment]
Have the 30c model and it keeps crashing my homebridge as well...was there any update on any potential fix?

**Comment by apexad** [comment]
Sems to only be the 30C model.  I have 15C and it's working fine.  Hard to troubleshoot with my vacuum since it's working, but I do have some ideas as to what might fix it based on the discussion here.  Hopefully I'll finally have some time this weekend to work on this.

**Comment by heyventrilo** [comment]
Is not a 30c problem, is an issue with the internet connection of the vacuum. If it get offline, it will crash the homebridge instance. I have made a separated bash  script to make it work as a separated process an no issues for months. I didnt have time to refactor the code but i will try to doit this week. Is something outside this plugin, but can help the people that have this issue.

**Comment by apexad** [comment]
@heyventrilo That makes sense.  Curious as to why my Vacuum is staying online when it appears it is disconnecting for others.  Anyway, look forward to your PR. Still have not found time to work on this myself.

**Comment by apexad** [comment]
`sudo npm install -g homebridge-eufy-robovac@2.1.0-beta.1`

Anyone want try this version and see if it's working?  I've added a `try...catch` block that I hope will correct what appears to be a stale connection issue.  It will re-establish the connection.

**Comment by apexad** [comment]
@heyventrilo checkout https://github.com/joshstrange/homebridge-eufy-robovac/commit/66bda47e5621a89ec472bb7b1bb5927c36fa13f8 for what I did.  Perhaps this may be what you were missing?  Essentially by using cmdswitch2 I assume you are creating a new connection every time, which is why it works. I'm hoping that this can be done in the plugin as well.  This is what I am attempting to do here.

**Comment by MonsterEnergy-wtf** [comment]
Installed beta and will let you all know what my experiences are. I still restarted the homebridge every 15 minutes due timeouts. 
Looking forward if this might resolve my issues too. 
I have three of 30Cs running.

**Comment by mateen110110** [comment]
Using latest beta and seems to still have that issue, although happening less frequently?

**Comment by apexad** [comment]
Question is does it totally crash Homebridge/Hoobs to where it cannot recover? If I've made progress to where it no longer crashes without a restart or uninstalling the plugin I can proceed and do what was mentioned previously.

The beta is essentially waiting for the device to fail/timeout. Once it does, it reconnects. What I may I've to do is auto-return Off while the reconnect happens or something like that. Also need to look into reducing the timeout as mentioned in this thread as well.

**Comment by MonsterEnergy-wtf** [comment]
So I can report that the issues still exist. In my case I get timeouts.
I also changed timout period in tuya api from 1 second to 5 and from my perspective it seems a bit more relaxed.
But in general the homebrigde crashes totally in my case. I have a temperature thing also running and after a crash both plugins are not working anymore. 
Even the logger to the console crashes so I have to restart homebridge service. When I now take a look into the backend I can see the crash dumps (which were logged but not displayed due the backend of homebridge crashed.

To become a bit more strange: as the plugin crashes it seems to be a "zombie" process which totally blocks homebrigde to work. 
homebridge.service: Found left-over process 5159 (node) in control group while starting unit. Ignoring.

After a full server restart everything is working fine until it crashes again. Really strange.

I highly suggest to add more try catches to catch all issues related to failed promises. For the beta you could output everything to the console too.

**Comment by apexad** [comment]
I've pushed beta 2.
`sudo npm install -g homebridge-eufy-robovac@2.1.0-beta.2`
All of the **get** functions are now wrapped in `try...catch` blocks.  All of them will return the `callback` immediately except for the main `getCleanState` which is what run **setup** function again to re-connect the instance.

**Comment by apexad** [comment]
I hope most of this is now resolved with v2.1.0
If this is not the case, re-open the issue.

**Comment by mreassassin** [comment]
I'm getting this same error.

My 15c almost always dies before it makes it back to the docking station. It seems like when that happens it makes home kit unresponsive. 

I have two 15c's and if/when even one of them goes offline or errors it seems that home kit goes unresponsive. 

I've tried latest and beta 2 with the same results.
